pkg/migration/migrations/retry_jobs_with_exponential_backoff.go

```diff
@@ -0,0 +1,271 @@
+// Copyright 2021 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package migrations
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/cockroachdb/cockroach/pkg/clusterversion"
+	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
+	"github.com/cockroachdb/cockroach/pkg/keys"
+	"github.com/cockroachdb/cockroach/pkg/kv"
+	"github.com/cockroachdb/cockroach/pkg/migration"
+	"github.com/cockroachdb/cockroach/pkg/security"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descs"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/systemschema"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sessiondata"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
+	"github.com/cockroachdb/errors"
+	"github.com/kr/pretty"
+)
+
+// Target schema changes in the system.jobs table, adding two columns and an
+// index for job retries with exponential backoff functionality.
+const (
+	addColsQuery = `
+ALTER TABLE system.jobs
+  ADD COLUMN num_runs INT8 FAMILY claim, 
+  ADD COLUMN last_run TIMESTAMP FAMILY claim`
+	addIndexQuery = `
+CREATE INDEX jobs_run_stats_idx
+		ON system.jobs (claim_session_id, status, created)
+		STORING (last_run, num_runs, claim_instance_id)
+		WHERE ` + systemschema.JobsRunStatsIdxPredicate
```

Comment:
I'd love to see a test in the `jobs` package that does an `EXPLAIN` for all of the relevant jobs queries and makes sure that they use this index.